rangeserver/src/prefetching_buffer.rs

```diff
@@ -0,0 +1,968 @@
+                }
+            }
+        }
+    }
+
+    pub async fn fetch_failed(&self, key: Bytes, fetch_sequence_number: u64) {
+        let mut cur_state = self.state.lock().await;
+        // The fetch failed, so change from Loading back to Requested
+        cur_state
+            .key_state
+            .insert(key.clone(), KeyState::Requested(fetch_sequence_number));
+        if let Some(sender) = cur_state.key_state_sender.get(&key) {
+            let _ = sender.send(KeyState::Requested(fetch_sequence_number));
+        }
+    }
+
+    /// Marks a transaction as being complete by updating bookkeeping to remove
+    /// the request from the data structures
+    pub async fn process_transaction_complete(&self, transaction_id: Uuid) {
+        let mut cur_state = self.state.lock().await;
+        let keys = cur_state.transaction_keys.get(&transaction_id).cloned();
+
+        // Loop over set of transaction keys to call remove key transactions (accept only transaction id)
+        if let Some(keys) = keys {
+            for key in keys {
+                // Log that this key is no longer being requested by this transaction
+                self.remove_from_key_transactions(
+                    &mut cur_state.key_transactions,
+                    transaction_id,
+                    key.clone(),
+                );
+                // If the key is no longer requested by any transaction, also delete it from the BTree
+                if !cur_state.key_transactions.contains_key(&key) {
+                    self.transaction_delete(key.clone(), &mut cur_state).await;
+                }
+            }
+        }
+        // Remove transaction from transaction_key
+        let _ = cur_state.transaction_keys.remove(&transaction_id);
+    }
+
+    /// Update the prefetch_store and update the key_state
+    pub async fn upsert(&self, key: Bytes, value: Bytes) {
+        let mut cur_state = self.state.lock().await;
+        // If key is still being requested by a different transaction, update BTree with latest value
+        if cur_state.key_transactions.contains_key(&key.clone()) {
+            let _ = cur_state.prefetch_store.insert(key.clone(), value);
+            // If the key is still loading in a prefetch, change to fetched
+            if let KeyState::Loading(_) = cur_state.key_state.get(&key).unwrap() {
+                cur_state.key_state.insert(key.clone(), KeyState::Fetched);
+                // Notify all watchers of the state change
+                if let Some(sender) = cur_state.key_state_sender.get(&key) {
+                    let _ = sender.send(KeyState::Fetched);
+                }
+            }
+        } else {
+            // If key is not being requested by any transactions, remove from prefetch_store instead of update
+            let _ = cur_state.prefetch_store.remove(&key); // might be None
+            let _ = cur_state.key_state.remove(&key); // might be None
```

Comment:
you already have a function that does these two lines. Also, I don't think doing it during upsert is very useful.
I would change the structure of this function slightly: check if the key is in the key_state map (in whatever state), then do the transition to Fetched and notification, and don't worry about evicting it here since it's done after the transaction completes.